Kernel/Net/TCPSocket.cpp

```diff
@@ -414,19 +415,45 @@ NetworkOrdered<u16> TCPSocket::compute_tcp_checksum(IPv4Address const& source, I
 
 ErrorOr<void> TCPSocket::protocol_bind()
 {
-    return m_adapter.with([this](auto& adapter) -> ErrorOr<void> {
+    dbgln_if(TCP_SOCKET_DEBUG, "TCPSocket::protocol_bind(), local_port() is {}", local_port());
+    // Check that we do have the address we're trying to bind to.
+    TRY(m_adapter.with([this](auto& adapter) -> ErrorOr<void> {
         if (has_specific_local_address() && !adapter) {
             adapter = NetworkingManagement::the().from_ipv4_address(local_address());
             if (!adapter)
                 return set_so_error(EADDRNOTAVAIL);
         }
         return {};
-    });
-}
-
-ErrorOr<void> TCPSocket::protocol_listen(bool did_allocate_port)
-{
-    if (!did_allocate_port) {
+    }));
+
+    if (local_port() == 0) {
+        // Allocate an unused ephemeral port.
+        constexpr u16 first_ephemeral_port = 32768;
+        constexpr u16 last_ephemeral_port = 60999;
+        constexpr u16 ephemeral_port_range_size = last_ephemeral_port - first_ephemeral_port;
+        u16 first_scan_port = first_ephemeral_port + get_good_random<u16>() % ephemeral_port_range_size;
+
+        return sockets_by_tuple().with_exclusive([&](auto& table) -> ErrorOr<void> {
+            for (u16 port = first_scan_port;;) {
+                IPv4SocketTuple proposed_tuple(local_address(), port, peer_address(), peer_port());
+
+                auto it = table.find(proposed_tuple);
```

Comment:
Nit: If the socket table is ordered by local address and port (which I'd guess it is?), we could reuse one iterator and improve performance a little over searching (up to) the entire table each time, but it's probably not significant.